File: pcc/py_runtime/src/py_int_addsub.c
/* pcc/py_runtime/src/py_int_addsub.c
 *
 * Bignum add/sub split from py_int.c for independent pcc-Python
 * replacement. Multiplication stays in py_int.c for now because its
 * uint32*uint32 intermediate needs full unsigned 64-bit behavior.
 */

#include "py_internal.h"
#include <stdlib.h>
#include <stdint.h>

static void py_bigint_normalize(PyIntObject *b) {
    while (b->ndigits > 0 && b->digits[b->ndigits - 1] == 0) {
        b->ndigits--;
    }
    if (b->ndigits == 0) b->sign = 0;
    else if (b->sign == 0) b->sign = 1;
}

static PyIntObject *bigint_copy(const PyIntObject *a) {
    PyIntObject *r = py_bigint_alloc(a->ndigits);
    if (r == NULL) return NULL;
    r->sign = a->sign;
    for (int32_t i = 0; i < a->ndigits; i++) r->digits[i] = a->digits[i];
    return r;
}

static int bigint_abs_cmp(const PyIntObject *a, const PyIntObject *b) {
    if (a->ndigits != b->ndigits) return a->ndigits < b->ndigits ? -1 : 1;
    for (int32_t i = a->ndigits - 1; i >= 0; i--) {
        if (a->digits[i] != b->digits[i]) {
            return a->digits[i] < b->digits[i] ? -1 : 1;
        }
    }
    return 0;
}

/* r = |a| + |b|. Sign of r is `sign_if_nonzero` when the sum is nonzero. */
static PyIntObject *bigint_abs_add(const PyIntObject *a, const PyIntObject *b,
                                   int32_t sign_if_nonzero) {
    int32_t la = a->ndigits, lb = b->ndigits;
    int32_t lr = (la > lb ? la : lb) + 1;
    PyIntObject *r = py_bigint_alloc(lr);
    if (r == NULL) return NULL;
    uint64_t carry = 0;
    for (int32_t i = 0; i < lr; i++) {
        uint64_t av = (i < la) ? a->digits[i] : 0;
        uint64_t bv = (i < lb) ? b->digits[i] : 0;
        uint64_t sum = av + bv + carry;
        r->digits[i] = (uint32_t)(sum & 0xFFFFFFFFu);
        carry = sum >> 32;
    }
    r->sign = sign_if_nonzero;
    py_bigint_normalize(r);
    return r;
}

/* r = |a| - |b|, requires |a| >= |b|. */
static PyIntObject *bigint_abs_sub(const PyIntObject *a, const PyIntObject *b,
                                   int32_t sign_if_nonzero) {
    int32_t la = a->ndigits, lb = b->ndigits;
    PyIntObject *r = py_bigint_alloc(la);
    if (r == NULL) return NULL;
    int64_t borrow = 0;
    for (int32_t i = 0; i < la; i++) {
        int64_t av = (int64_t)(uint64_t)a->digits[i];
        int64_t bv = (i < lb) ? (int64_t)(uint64_t)b->digits[i] : 0;
        int64_t diff = av - bv - borrow;
        if (diff < 0) { diff += ((int64_t)1 << 32); borrow = 1; }
        else          { borrow = 0; }
        r->digits[i] = (uint32_t)(diff & 0xFFFFFFFFu);
    }
    r->sign = sign_if_nonzero;
    py_bigint_normalize(r);
    return r;
}

PyIntObject *py_bigint_add(const PyIntObject *a, const PyIntObject *b) {
    if (a->sign == 0) return bigint_copy(b);
    if (b->sign == 0) return bigint_copy(a);
    if (a->sign == b->sign) return bigint_abs_add(a, b, a->sign);
    /* Opposite signs: subtract smaller magnitude from larger. */
    int c = bigint_abs_cmp(a, b);
    if (c == 0) {
        PyIntObject *r = py_bigint_alloc(0);
        return r;   /* zero */
    }
    if (c > 0) return bigint_abs_sub(a, b, a->sign);
    return bigint_abs_sub(b, a, b->sign);
}

PyIntObject *py_bigint_sub(const PyIntObject *a, const PyIntObject *b) {
    if (b->sign == 0) return bigint_copy(a);
    if (a->sign == 0) {
        PyIntObject *r = bigint_copy(b);
        if (r == NULL) return NULL;
        r->sign = -r->sign;
        return r;
    }
    if (a->sign != b->sign) {
        /* a - b with opposite signs == addition of magnitudes, sign of a. */
        return bigint_abs_add(a, b, a->sign);
    }
    /* Same sign: |a| - |b|. */
    int c = bigint_abs_cmp(a, b);
    if (c == 0) return py_bigint_alloc(0);
    if (c > 0) return bigint_abs_sub(a, b, a->sign);
    return bigint_abs_sub(b, a, -a->sign);
}

File: pcc/py_runtime/py/py_int_addsub.py
"""pcc-Python replacement for py_runtime/src/py_int_addsub.c."""

__pcc_runtime_port__ = True

from pcc.extern import extern, c_abi_export, c_int64, c_ptr
from pcc.py_runtime.py.py_abi_constants import (
    PYINTOBJECT_DIGITS_OFFSET,
    PYINTOBJECT_NDIGITS_OFFSET,
    PYINTOBJECT_SIGN_OFFSET,
)
from pcc.unsafe import load_i32, ptr_is_null, store_i32


py_bigint_alloc = extern("py_bigint_alloc", (c_int64,), c_ptr)


def _load_u32(obj, offset: int) -> int:
    v: int = load_i32(obj, offset)
    if v < 0:
        v = v + 4294967296
    return v


def _store_u32(obj, offset: int, value: int) -> None:
    store_i32(obj, offset, value)


def _normalize(b) -> None:
    ndigits: int = load_i32(b, PYINTOBJECT_NDIGITS_OFFSET)
    while ndigits > 0 and _load_u32(b, PYINTOBJECT_DIGITS_OFFSET + (ndigits - 1) * 4) == 0:
        ndigits = ndigits - 1
    store_i32(b, PYINTOBJECT_NDIGITS_OFFSET, ndigits)
    if ndigits == 0:
        store_i32(b, PYINTOBJECT_SIGN_OFFSET, 0)
    elif load_i32(b, PYINTOBJECT_SIGN_OFFSET) == 0:
        store_i32(b, PYINTOBJECT_SIGN_OFFSET, 1)


def _bigint_copy(a):
    ndigits: int = load_i32(a, PYINTOBJECT_NDIGITS_OFFSET)
    r = py_bigint_alloc(ndigits)
    if ptr_is_null(r):
        return r
    store_i32(r, PYINTOBJECT_SIGN_OFFSET, load_i32(a, PYINTOBJECT_SIGN_OFFSET))
    i: int = 0
    while i < ndigits:
        store_i32(r, PYINTOBJECT_DIGITS_OFFSET + i * 4, load_i32(a, PYINTOBJECT_DIGITS_OFFSET + i * 4))
        i = i + 1
    return r


def _abs_cmp(a, b) -> int:
    na: int = load_i32(a, PYINTOBJECT_NDIGITS_OFFSET)
    nb: int = load_i32(b, PYINTOBJECT_NDIGITS_OFFSET)
    if na != nb:
        if na < nb:
            return -1
        return 1
    i: int = na - 1
    while i >= 0:
        av: int = _load_u32(a, PYINTOBJECT_DIGITS_OFFSET + i * 4)
        bv: int = _load_u32(b, PYINTOBJECT_DIGITS_OFFSET + i * 4)
        if av != bv:
            if av < bv:
                return -1
            return 1
        i = i - 1
    return 0


def _abs_add(a, b, sign_if_nonzero: int):
    la: int = load_i32(a, PYINTOBJECT_NDIGITS_OFFSET)
    lb: int = load_i32(b, PYINTOBJECT_NDIGITS_OFFSET)
    lr: int = lb + 1
    if la > lb:
        lr = la + 1
    r = py_bigint_alloc(lr)
    if ptr_is_null(r):
        return r
    carry: int = 0
    i: int = 0
    while i < lr:
        av: int = 0
        bv: int = 0
        if i < la:
            av = _load_u32(a, PYINTOBJECT_DIGITS_OFFSET + i * 4)
        if i < lb:
            bv = _load_u32(b, PYINTOBJECT_DIGITS_OFFSET + i * 4)
        total: int = av + bv + carry
        _store_u32(r, PYINTOBJECT_DIGITS_OFFSET + i * 4, total & 4294967295)
        carry = total >> 32
        i = i + 1
    store_i32(r, PYINTOBJECT_SIGN_OFFSET, sign_if_nonzero)
    _normalize(r)
    return r


def _abs_sub(a, b, sign_if_nonzero: int):
    la: int = load_i32(a, PYINTOBJECT_NDIGITS_OFFSET)
    lb: int = load_i32(b, PYINTOBJECT_NDIGITS_OFFSET)
    r = py_bigint_alloc(la)
    if ptr_is_null(r):
        return r
    borrow: int = 0
    i: int = 0
    while i < la:
        av: int = _load_u32(a, PYINTOBJECT_DIGITS_OFFSET + i * 4)
        bv: int = 0
        if i < lb:
            bv = _load_u32(b, PYINTOBJECT_DIGITS_OFFSET + i * 4)
        diff: int = av - bv - borrow
        if diff < 0:
            diff = diff + 4294967296
            borrow = 1
        else:
            borrow = 0
        _store_u32(r, PYINTOBJECT_DIGITS_OFFSET + i * 4, diff & 4294967295)
        i = i + 1
    store_i32(r, PYINTOBJECT_SIGN_OFFSET, sign_if_nonzero)
    _normalize(r)
    return r


@c_abi_export("py_bigint_add")
def py_bigint_add(a, b):
    sa: int = load_i32(a, PYINTOBJECT_SIGN_OFFSET)
    sb: int = load_i32(b, PYINTOBJECT_SIGN_OFFSET)
    if sa == 0:
        return _bigint_copy(b)
    if sb == 0:
        return _bigint_copy(a)
    if sa == sb:
        return _abs_add(a, b, sa)
    c: int = _abs_cmp(a, b)
    if c == 0:
        return py_bigint_alloc(0)
    if c > 0:
        return _abs_sub(a, b, sa)
    return _abs_sub(b, a, sb)


@c_abi_export("py_bigint_sub")
def py_bigint_sub(a, b):
    sa: int = load_i32(a, PYINTOBJECT_SIGN_OFFSET)
    sb: int = load_i32(b, PYINTOBJECT_SIGN_OFFSET)
    if sb == 0:
        return _bigint_copy(a)
    if sa == 0:
        r = _bigint_copy(b)
        if ptr_is_null(r):
            return r
        store_i32(r, PYINTOBJECT_SIGN_OFFSET, 0 - load_i32(r, PYINTOBJECT_SIGN_OFFSET))
        return r
    if sa != sb:
        return _abs_add(a, b, sa)
    c: int = _abs_cmp(a, b)
    if c == 0:
        return py_bigint_alloc(0)
    if c > 0:
        return _abs_sub(a, b, sa)
    return _abs_sub(b, a, 0 - sa)
